File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.py
import sys
import heapq
input = sys.stdin.readline

def dijkstra(start, n, graph):
    # 최소 비용을 저장할 배열
    distances = [float('inf')] * (n + 1)
    distances[start] = 0
    pq = [(0, start)]  # (비용, 노드)

    while pq:
        current_cost, current_node = heapq.heappop(pq)
        
        # 이미 처리된 노드라면 건너뛴다
        if distances[current_node] < current_cost:
            continue
        
        # 인접 노드를 탐색
        for neighbor, cost in graph[current_node]:
            new_cost = current_cost + cost
            if new_cost < distances[neighbor]:
                distances[neighbor] = new_cost
                heapq.heappush(pq, (new_cost, neighbor))
    
    return distances

# 입력 받기
n = int(input())
m = int(input())
graph = {i: [] for i in range(1, n + 1)}

for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))

start, end = map(int, input().split())

# 다익스트라 알고리즘 실행
distances = dijkstra(start, n, graph)

# 결과 출력
print(distances[end])

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node implements Comparable<Node> {
        int node;
        int cost;
        
        public Node(int node, int cost) {
            this.node = node;
            this.cost = cost;
        }
        
        @Override
        public int compareTo(Node o) {
            return this.cost - o.cost;
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        ArrayList<ArrayList<Node>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            graph.get(a).add(new Node(b, c));
        }
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        int[] distances = dijkstra(start, n, graph);

        System.out.println(distances[end]);
        
        br.close();
    }
    
    public static int[] dijkstra(int start, int n, ArrayList<ArrayList<Node>> graph) {

        int[] distances = new int[n + 1];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;
        
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));  // (노드, 비용)
        
        while (!pq.isEmpty()) {
            Node current = pq.poll();
            int currentNode = current.node;
            int currentCost = current.cost;

            if (distances[currentNode] < currentCost) {
                continue;
            }

            for (Node neighbor : graph.get(currentNode)) {
                int newCost = currentCost + neighbor.cost;
                if (newCost < distances[neighbor.node]) {
                    distances[neighbor.node] = newCost;
                    pq.offer(new Node(neighbor.node, newCost));
                }
            }
        }
        
        return distances;
    }
}